### PR TITLE
Use an accumulator for the acceptance in the samplers

### PR DESCRIPTION
**Pull request by inailuig**
avoids doing a horizontal sum over all the chains at every sampling step.

### CODE REVIEW

File: netket/sampler/metropolis.py [hunk 1]
```diff
@@ -54,8 +54,10 @@ class MetropolisSamplerState(SamplerState):
         default_factory=lambda: jnp.zeros((), dtype=jnp.int64)
     )
     """Number of moves performed along the chains in this process since the last reset."""
-    n_accepted_proc: int = struct.field(
-        default_factory=lambda: jnp.zeros((), dtype=jnp.int64)
+    n_accepted_proc: jnp.ndarray = struct.field(
+        default_factory=lambda: jnp.zeros(
+            (), dtype=jnp.int64
+        )  # TODO (c): initialize with shape σ.shape[0] to avoid recompilation
```
Comment (inailuig) [inline netket/sampler/metropolis.py:60]: I assume I would have to use __post__init__ to infer the shape from \sigma?

File: netket/sampler/metropolis.py [hunk 2]
```diff
@@ -332,14 +338,14 @@ def loop_body(i, s):
             "σ": state.σ,
             "log_prob": sampler.machine_pow * machine.apply(parameters, state.σ).real,
             # for logging
-            "accepted": state.n_accepted_proc,
+            "accepted": jnp.zeros(state.σ.shape[0], dtype=jnp.int64),
```
Comment (inailuig) [inline netket/sampler/metropolis.py:341]: then I could revert this change

File: netket/sampler/metropolis.py (hunk 1 above)
Comment (PhilipVinc) [reply netket/sampler/metropolis.py:60]: yes. Either you initialise it with `None` and then set it to whatever you need,
or you use `pre_init` and add it to the kwargs.

File: netket/sampler/metropolis.py [hunk 3]
```diff
@@ -79,7 +81,7 @@ def n_steps(self) -> int:
     @property
     def n_accepted(self) -> int:
         """Total number of moves accepted across all processes since the last reset."""
-        res, _ = mpi.mpi_sum_jax(self.n_accepted_proc)
+        res, _ = mpi.mpi_sum_jax(jnp.sum(self.n_accepted_proc))
```
Comment (inailuig) [inline netket/sampler/metropolis.py:84]: the sum for the acceptance ratio is taken here, the same place were we already did it for mpi

Comment (PhilipVinc) [reply netket/sampler/metropolis.py:84]: good. so can you finalise this pr?

**Review by PhilipVinc — APPROVED** [review]
Looks good now!

**Comment by PhilipVinc** [comment]
@inailuig the test failure seems unrelated?

**Comment by inailuig** [comment]
seems indeed unrelated, but let's run it again anyway.

**Comment by PhilipVinc** [comment]
The test seems now broken also on master, so I'm merging anyway.
